core/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsContext.java

```diff
@@ -214,4 +253,38 @@ private boolean isChildHit(SearchHit hit) {
             return documentMapper.type().equals(hitDocumentMapper.parentFieldMapper().type());
         }
     }
+
+    static void intersect(Weight weight, Weight innerHitQueryWeight, Collector collector, LeafReaderContext ctx) throws IOException {
+        Scorer scorer = weight.scorer(ctx);
+        if (scorer == null) {
+            return;
+        }
```

Comment:
This can be made even more efficient with range queries by doing something like this:

```
ScorerSupplier scorerSupplier = weight.scorerSupplier(ctx);
if (scorerSupplier == null) {
  return;
}
Scorer scorer = scorerSupplier.get(true); // use random access since this scorer will be consumed on a minority of documents
```

(Note that you do not need to null check the scorer if you do that, ScorerSupplier.get is guaranteed to not return null.)